### PR TITLE
#66 - make static helpers replacable

### LINKED ISSUE #66: Make static helpers replaceable
I would like to provide a way to replace these static helpers that started to showing up in this package.

For example `UserHelper:getBy("email", "example@example.com)` is reaching to other class and doing some magic. What if someone would like to replace it with simple 
```php
    public static function getBy(string $field, string $value): ?object
    {
        return \MyApplication\Eloquent\User::query()->where($field, $value)->first();
    }
```
or maybe something more complex?

We could create a small service container for contexts and connect it to configuratinon file. So, for example, instead of using `UserHelper:getBy("email", "example@example.com)`, we could use `ContextHelper::get(UserHelper::class)->getBy("email", "example@example.com)` or even `ContextHelper::get("user")->getBy("email", "example@example.com)`

And then, config file could have an array of options than could be changed by end user:
```
"helpers" => [
    "user" => \Blumilk\BLT\Helpers\UserHelper::class,
    // (...)
],
```

Service provider should be empty-config-proof.

### PR DESCRIPTION
**Pull request by JakubKermes**
This should close #66.

### CODE REVIEW

File: config/blt.php [hunk 1]
```diff
@@ -1,10 +1,20 @@
...
     "namespaces" => [
         "default" => "App\\",
         "types" => [],
     ],
+    "helpers" => [
+        "array" => ArrayHelper::class,
+        "class" => ClassHelper::class,
+        "context" => ContextHelper::class,
+        "nullable" => NullableHelper::class,
+    ],
```
Comment (PiotrFedak) [inline config/blt.php:19]:
```suggestion
    "helpers" => [
        "array" => ArrayHelper::class,
        "class" => ClassHelper::class,
        "context" => ContextHelper::class,
        "nullable" => NullableHelper::class,
        "boolean" => BooleanHelper::class,
    ],
```

Comment (PiotrFedak) [reply config/blt.php:19]: You forgot one of the helpers

**Review by PiotrFedak — CHANGES_REQUESTED** [review]

**Review by PiotrFedak — DISMISSED** [review]

File: src/Helpers/ContextHelper.php [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+use Illuminate\Support\Str;
+
+class ContextHelper
+{
+    public static function getHelper(
+        string $helper,
+    ) {
+        $helperClass = config("blt.helpers.$helper") ?? "Blumilk\BLT\Helpers\\" . Str::ucfirst($helper) . "Helper";
+
+        return new $helperClass();
```
Comment (krzysztofrewak) [inline src/Helpers/ContextHelper.php:16]: I think I would avoid things like this. Maybe `return match($helper) (...)` would be more safe?

Comment (krzysztofrewak) [reply src/Helpers/ContextHelper.php:16]: With cases like `"user" => new UserHelper()` and so on.

Comment (krzysztofrewak) [reply src/Helpers/ContextHelper.php:16]: Or maybe, just maybe, add something like `::getUserHelper()`?

Comment (JakubKermes) [reply src/Helpers/ContextHelper.php]: Something like this?

File: src/Helpers/ContextHelper.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+            "array" => config("blt.helpers.array"),
+            "boolean" => config("blt.helpers.boolean"),
+            "class" => config("blt.helpers.class"),
+            "context" => config("blt.helpers.context"),
+            "nullable" => config("blt.helpers.nullable"),
+        };
+
+        if (!$helperClass) {
+            $helperClass = "Blumilk\BLT\Helpers\\" . Str::studly($helper) . "Helper";
+        }
```
Comment (krzysztofrewak) [inline src/Helpers/ContextHelper.php:23]:
```suggestion
            default => "Blumilk\BLT\Helpers\\" . Str::studly($helper) . "Helper";
        };
```

File: src/Helpers/ContextHelper.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+        if (!class_exists($helperClass)) {
+            throw new \Exception("Helper class $helperClass does not exist.");
```
Comment (krzysztofrewak) [inline src/Helpers/ContextHelper.php:26]: Please import this or maybe create your own?

File: config/blt.php [hunk 5]
```diff
@@ -1,10 +1,22 @@
...
 return [
     "namespaces" => [
         "default" => "App\\",
         "types" => [],
     ],
+    "helpers" => [
+        "array" => ArrayHelper::class,
+        "boolean" => BooleanHelper::class,
+        "class" => ClassHelper::class,
+        "context" => ContextHelper::class,
```
Comment (krzysztofrewak) [inline config/blt.php:19]: Do you think we will be able to replace context helper?

File: src/Helpers/ContextHelper.php [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        }
+
+        return new $helperClass();
```
Comment (krzysztofrewak) [inline src/Helpers/ContextHelper.php:29]: As far as I think about it, I think I prefer multiple method like `getUserHelper(): UserHelper` and so on. I don't see any case when someone would like to use non-existing helper here.

File: src/Helpers/ContextHelper.php [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+namespace Blumilk\BLT\Helpers;
+
+use Illuminate\Support\Str;
+
+class ContextHelper
+{
+    public static function getHelper(string $helper)
+    {
+        $helperClass = match($helper) {
+            "array" => config("blt.helpers.array"),
```
Comment (krzysztofrewak) [inline src/Helpers/ContextHelper.php:14]:
```suggestion
            "array" => config("blt.helpers.array") ?? ArrayHelper::class,
```

And please change all of them this way.

**Review by krzysztofrewak — CHANGES_REQUESTED** [review]

**Comment by EwelinaSkrzypacz** [comment]
Conflicts :(

File: src/Helpers/ContextHelper.php [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+    public static function getClassHelper(): ClassHelper
+    {
+        return new (config("blt.helpers.class") ?? ClassHelper::class);
+    }
+
+    public static function getNullableHelper(): NullableHelper
+    {
+        return new (config("blt.helpers.nullable") ?? NullableHelper::class);
+    }
```
Comment (krzysztofrewak) [inline src/Helpers/ContextHelper.php:27]:
```suggestion
    public static function getArrayHelper(): ArrayHelper
    {
        return new (config("blt.helpers.array") ?? ArrayHelper::class)();
    }

    public static function getBooleanHelper(): BooleanHelper
    {
        return new (config("blt.helpers.boolean") ?? BooleanHelper::class)();
    }

    public static function getClassHelper(): ClassHelper
    {
        return new (config("blt.helpers.class") ?? ClassHelper::class)();
    }

    public static function getNullableHelper(): NullableHelper
    {
        return new (config("blt.helpers.nullable") ?? NullableHelper::class)();
    }
```

File: config/blt.php [hunk 9]
```diff
@@ -9,4 +13,10 @@
             "User" => "App\Models\User",
         ],
     ],
+    "helpers" => [
+        "array" => ArrayHelper::class,
+        "boolean" => BooleanHelper::class,
+        "class" => ClassHelper::class,
+        "nullable" => NullableHelper::class,
```
Comment (krzysztofrewak) [inline config/blt.php:20]: What about User helper? I think it's the most obvious thing to overwrite here.

**Review by krzysztofrewak — CHANGES_REQUESTED** [review]

**Review by krzysztofrewak — APPROVED** [review]

**Review by PiotrFedak — APPROVED** [review]

**Pull request merged**